### ISSUE TITLE
Detect Programatic change.

### ISSUE THREAD

**Issue by paulsUsername** [chicio/RangeUISlider#52]
Programmatic calling of rangeChangeFinished is causing a lot of issues for me. Is there any way to detect if it is from a human or not? 

I see the programmaticChangeCompleted method getting called internally to RangeUISlider. This could easily pass a "isFromUser" bool to make it easier to discern, "real" changes and merely the range slider initialising itself.

**Comment by chicio** [comment]
ciao @paulsUsername, can you send me some code example of the issue? which kind of problem do you have?
To be honest I don't understand why you need to call the rangeChangeFinished delegate method directly: this is called by RangeUISlider directly and you just have to declare it in a delegate of our RangeUISlider delegate.

Thanks,
Fabrizio

**Comment by ppbdrinker** [comment]
If I understood correctly, what @paulsUsername describes here is that programmatic changes to the slider's position should not result in the same callback as the human interaction with the UI. 

E.g. think of a `UITextField`:  updating its' `text` property does not fire a `textDidChange` delegate callback.

Currently this makes the two-way view model bindings impossible, i.e. an infinite loop is triggered, if slider is glued to the changes in the underlying viewmodel, and the delegate callbacks are in turn used to update values in the viewmodel itself.

Once you make a call to the `changeLeftKnob`, the slider calls the `delegate.rangeChangeFinished` callback, where an update to the view model is peformed. The MVVM binding, in turn, leads to the next iteration in this loop.

A workaround is to break the loop if the `minValueSelected` parameter of the delegate callback is the same as the current one, but since those are floats, it's a so-so approach.

**Comment by chicio** [comment]
Hi @ppbdrinker , ok now I understood the problem (and it makes sense, especially in case of two way bindings with combine/Rx frameworks). I can try to add some callback to the delegate that are called specifically for the programmatic changes. In this way you can understand the source of changes and adjust your logic consequently. What do you think?

**Comment by ppbdrinker** [comment]
Hey @chicio, thanks for taking a look into it, much appreciated!

I think it should be vice versa, and the callback for user interaction changes should be modified (either to include a separate call parameter, like @paulsUsername suggests, or the callback shouldn't be called upon programmatic changes at all).

Having a dedicated callback to track the programmatic changes won't solve the issue: those are the changes that are in fact undesired to be tracked (and the developer knows about them already, since it is the developer who writes the code to perform those programmatic changes).

If the UI-induced changes callback stays as it is, I believe having an additional one for the programmatic changes won't help much for the case we are discussing.

iow, it is the _user input_ that makes the undeterministic part of the dataflow. It cannot be predicted, and thus requires notifications/delegate calls when it occurs.
Algorithmic changes are determined, and the application code is aware of them _before_ they are applied, so tracking them is a nobrainer, and no callback is needed.

let me know if this makes sense please!

**Comment by chicio** [comment]
Hi @ppbdrinker , makes sense. I will try to develop this stuff in the next week(ends 😂).

Thanks,
Fabrizio.

**Comment by chicio** [comment]
Hi @paulsUsername @ppbdrinker , RangeUISlider 4.0.0 release. This is a major breaking change because in the end I changed the response of the delegate with the additional parameter to detect user interaction vs programmatic changes. I chose this way because I didn't find another way to test property the core parts of the library (in particular the programmatic changes).

Thank you for you patience and support and remember to star my library here on GitHub ⭐
